### ISSUE TITLE
OpenTelemetryLayer implement Filtered

### ISSUE THREAD

**Issue by BrynCooke** [tokio-rs/tracing-opentelemetry#15]
### Motivation

It seems that `OpenTelemetryLayer` may be able to make use of `Filtered` to allow `Registry` to avoid creating spans in some cases.

For the vast majority of cases users will set their otel sampling rate to something low, but this doesn't have an effect on the tracing side of things, and seems to have quite a performance hit.

Otel is complicated, and I am very aware that I may have missed the reason as to why this is not feasible while remaining compatible with the otel spec. 

### Proposal

Add a new function `OpenTelemetryLayer::prefiltered(self)` that returns the layer wrapped via `with_filter`.
Move the logic for deciding if a span is of interest into the Filter. Only apply the logic for root spans.

The downside is that the tracer would have to be moved to an `Arc` so that it can be shared between the filter function and the layer.

### Alternatives
To avoid the arc possibly modify the `FilterLayer` to allow self filtering if a layer implements `Filtered`.

**Comment by davidbarsky** [comment]
Hmm, I'm not entirely sure why this _strictly_ needed as opposed to writing `tracing_opentelemetry::layer().with_filter(filter).with_tracer(tracer)`.

(I'm genuinely unsure, so I'd love to get clarification!)

**Comment by BrynCooke** [comment]
I'm not really explaining this well.

I'm assuming that we are on the same page regarding the intent, that the filter would take on the work that is currently inside the otel layer to call `sampled_context`. Calling of `sampled_context` would no longer be in the remit of the otel layer itself.

Then from the user point of view I'd like to be able to call:
`tracing_opentelemetry::layer().with_tracer(tracer)`
And have it internally sort out the details of wrapping the otel layer in a filter for performance.

However this would probably be a breaking change, so maybe some different api or `prefiltered` method on otel layer could be in order.

**Comment by BrynCooke** [comment]
Looks like https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html#method.on_layer can be used to add the filter so no new APIs would be needed.

**Comment by 0xForerunner** [comment]
Just coming across this now.
```rust
        let filter = Targets::new()
            .with_default(LevelFilter::OFF)
            .with_target("my-crate", args.log_level);
        let layer = OpenTelemetryLayer::new(tracer).with_filter(filter);
```
doesn't seem to work for me at all? No matter what filter I use I am still getting spans from dependencies...

Is there a workaround for this? I'm assuming it's to create a sampler?

**Comment by mladedav** [comment]
@BrynCooke  I don't think this is currently possible since the sampling decision may happen later (or rather be changed) when `set_parent` is called. For example server spans when a new request arrives are created as a root span as far as `tracing` is concerned, but are then linked to a remote span after the fact. I don't think we should filter these out.

Similarly, I fairly often use a trick (and I've seen other people use it) where you manually break `tracing` parent/child relationships and then manually `set_parent` just for OpenTelemetry so that children do not keep parents open. These spans could also be mistakenly filtered out.

We might add a magic field telling the layer to not eagerly filter out the span, but it would be easy to forget to set this and then some spans inside traces will just disappear with no errors.

---

@0xForerunner This works for me, if you still have issues and want some help, open a new issue please.
